In [1]:
import numpy as np
import pandas as pd

In [5]:
df = pd.read_csv("./train.csv", usecols=['X1', 'X2', 'X3', 'X4', 'X5', 'X6'])

In [6]:
df.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [8]:
for col in df.columns:
    print(col, ': ', df[col].nunique(), 'unique labels')    

X1 :  27 unique labels
X2 :  44 unique labels
X3 :  7 unique labels
X4 :  4 unique labels
X5 :  29 unique labels
X6 :  12 unique labels


In [9]:
df.shape

(4209, 6)

In [10]:
#checking how many columns we willl have after one hot encoding the variables
pd.get_dummies(df, drop_first=True).shape

(4209, 117)

117 columns are being created for the 6 columns in the df

### How do we prevent this from happening? ###
limit one hot encoding to the 10 most frequent labels of the variable (column). Meaning, only make one binary variable for each of the 10 most frequent labels only.

In [15]:
top_10 = df.X2.value_counts().sort_values(ascending= False).head(10) #highest frequency values in descending order of column X2

In [17]:
top_10 = [i for i in top_10.index]

In [18]:
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [23]:
for label in top_10:
    df[label] = np.where(df["X2"] == label , 1, 0) #wherever X2 matches with label, that is top 10 label, put 1 else 0

In [24]:
df[["X2"]+top_10].head(20)

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0

In [25]:
def one_hot_top_X(df, variable, top_x_labels):
    for label in top_x_labels:
        df[variable+"_"+label] = np.where(df["X2"] == label , 1, 0)

In [26]:
df = pd.read_csv("./train.csv", usecols=['X1', 'X2', 'X3', 'X4', 'X5', 'X6'])

In [27]:
one_hot_top_X(df, 'X2', top_10)

In [28]:
df.head()

X1  X2 X3 X4 X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0  v  at  a  d  u  j      0      0      0     0      0     0     0     0   
1  t  av  e  d  y  l      0      0      0     0      0     0     0     0   
2  w   n  c  d  x  j      0      0      0     0      0     0     1     0   
3  t   n  f  d  x  l      0      0      0     0      0     0     1     0   
4  v   n  f  d  h  d      0      0      0     0      0     0     1     0   

   X2_f  X2_e  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0